README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# surbayes

<!-- badges: start -->

<!-- badges: end -->

The goal of surbayes is to provide tools for Bayesian analysis of the
seemingly unrelated regression (SUR) model. In particular, we implement
the direct Monte Carlo (DMC) approach of Zellner and Ando (2010). We
also implement a Gibbs sampler to sample from a power prior on the SUR
model.

## Installation

You can install the released version of surbayes from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("surbayes")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("ethan-alt/surbayes")
```

## Example

This is a basic example which shows you how to sample from the posterior

``` r
library(surbayes)
## Taken from bayesm package
M = 10 ## number of samples
set.seed(66)
## simulate data from SUR
beta1 = c(1,2)
beta2 = c(1,-1,-2)
nobs = 100
nreg = 2
iota = c(rep(1, nobs))
X1 = cbind(iota, runif(nobs))
X2 = cbind(iota, runif(nobs), runif(nobs))
Sigma = matrix(c(0.5, 0.2, 0.2, 0.5), ncol = 2)
U = chol(Sigma)
E = matrix( rnorm( 2 * nobs ), ncol = 2) %*% U
y1 = X1 %*% beta1 + E[,1]
y2 = X2 %*% beta2 + E[,2]
X1 = X1[, -1]
X2 = X2[, -1]
data = data.frame(y1, y2, X1, X2)
names(data) = c( paste0( 'y', 1:2 ), paste0('x', 1:(ncol(data) - 2) ))
## run DMC sampler
formula.list = list(y1 ~ x1, y2 ~ x2 + x3)

## Fit models
out_dmc = sur_sample( formula.list, data, M = M )            ## DMC used
#> Direct Monte Carlo sampling used
out_powerprior = sur_sample( formula.list, data, M, data )   ## Gibbs used
#> Gibbs sampling used for power prior model
```


NEWS.md
# surbayes 0.1.2

* Updated makevars file.

R/test.R
#' @useDynLib test
#' @importFrom Rcpp evalCpp
#' @exportPattern "^[[:alpha:]]+"
NULL

R/sur_sample.R
#' Sample from seemingly unrelated regression
#'
#' This function is a wrapper function that performs either (1) Direct Monte Carlo or (2) Gibbs sampling
#' of the SUR model depending on whether 1 or 2 data sets are specified.
#' 
#' @include sur_sample_dmc.R
#' @include sur_sample_powerprior.R
#' @importFrom stats as.formula lm model.matrix
#' 
#' @param formula.list A list of formulas, each element giving the formula for the corresponding endpoint.
#' @param data A \code{data.frame} containing all the variables contained in \code{formula.list]}
#' @param M Number of samples to be drawn
#' @param histdata A \code{data.frame} of historical data to apply power prior on
#' @param Sigma0 \emph{optional} a \eqn{J \times J} \code{matrix} giving the initial covariance matrix. Default is the MLE. Ignored if \code{histdata} is null
#' @param a0 A scalar between 0 and 1 giving the power prior parameter. Ignored if \code{histdata} is null
#' @param burnin A non-negative integer giving the burn-in parameter. Ignored if \code{histdata} is null as direct Monte Carlo is performed
#' @param thin A positive integer giving the thin parameter. Ignored if \code{histdata} is null as direct Monte Carlo is performed
#' @return A list. First element is posterior draws. Second element is list of JxJ covariance matrices.
#' 
#' @examples
#' ## Taken from bayesm package
#' if(nchar(Sys.getenv("LONG_TEST")) != 0) {M=1000} else {M=10}
#' set.seed(66)
#' ## simulate data from SUR
#' beta1 = c(1,2)
#' beta2 = c(1,-1,-2)
#' nobs = 100
#' nreg = 2
#' iota = c(rep(1, nobs))
#' X1 = cbind(iota, runif(nobs))
#' X2 = cbind(iota, runif(nobs), runif(nobs))
#' Sigma = matrix(c(0.5, 0.2, 0.2, 0.5), ncol = 2)
#' U = chol(Sigma)
#' E = matrix( rnorm( 2 * nobs ), ncol = 2) %*% U
#' y1 = X1 %*% beta1 + E[,1]
#' y2 = X2 %*% beta2 + E[,2]
#' X1 = X1[, -1]
#' X2 = X2[, -1]
#' data = data.frame(y1, y2, X1, X2)
#' names(data) = c( paste0( 'y', 1:2 ), paste0('x', 1:(ncol(data) - 2) ))
#' ## run DMC sampler
#' formula.list = list(y1 ~ x1, y2 ~ x2 + x3)
#' 
#' ## Fit models
#' out_dmc = sur_sample( formula.list, data, M = M )            ## DMC used
#' out_powerprior = sur_sample( formula.list, data, M, data )   ## Gibbs used
#' @export
sur_sample <- function(
  formula.list, 
  data, 
  M, 
  histdata = NULL,
  Sigma0 = NULL, 
  a0 = 1, 
  burnin = 0, 
  thin = 1
) {
  
  if ( is.null( histdata ) ) {
    message('Direct Monte Carlo sampling used')
    return ( sur_sample_dmc( formula.list, data, M ) )
  } else {
    message('Gibbs sampling used for power prior model')
    return ( sur_sample_powerprior( formula.list, data, histdata, M, Sigma0, a0, burnin, thin ) )
  }
}



R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' Fast kronecker product with response vector
#' 
#' This is a c++ implementation of the fast kronecker product with response vector
#'
#' @param Sigma covariance matrix
#' @param Y matrix of response variables (Y1, ..., YJ)
#' @param n number of observations
#' @param J number of endpoints
#' @return Returns a vector with result of \code{ kron(Sigma, diag(n)) \% y }
#' @keywords internal
fastKronEye_Y <- function(Sigma, Y, n, J) {
    .Call('_test_fastKronEye_Y', PACKAGE = 'test', Sigma, Y, n, J)
}

#' Fast kronecker product of crossproduct matrix
#' 
#' This is a c++ implementation of the fast kronecker product
#' t(X) %*% kron(Sigma, I) %*% X. It avoids computing the kronecker product
#'
#' @param XtX a matrix that is crossprod((X1, ..., XJ)) in R
#' @param Sigma JxJ covariance matrix
#' @param pvec J-dimensional vector giving number of observations for each endpoint
#' @param n number of observations
#' @param J number of endpoints
#' @keywords internal
#' @return \code{matrix} result of \eqn{X' (\Sigma \otimes I_n) X}
fastKronEye_crossprod <- function(XtX, Sigma, pvec, n, J) {
    .Call('_test_fastKronEye_crossprod', PACKAGE = 'test', XtX, Sigma, pvec, n, J)
}

#' Get one sample from predictive posterior of SUR
#' 
#' C++ implementation to obtain one sample from predictive posterior
#' density
#'
#' @param mu vector of means
#' @param Sigma covariance matrix shared among all observations
#' @param n number of observations
#' @param J number of endpoints
predict_surbayes_helper <- function(mu, Sigma, n, J) {
    .Call('_test_predict_surbayes_helper', PACKAGE = 'test', mu, Sigma, n, J)
}

#' Sample from predictive posterior density C++ helper
#' 
#' C++ implementation to obtain a matrix of samples from predictive posterior density
#'
#' @param Mu matrix of means
#' @param Sigmalist list of covariance matrices
#' @param n number of observations
#' @param J number of endpoints
#' @param nsims Number of simulations (number of rows in Mu)
predict_surbayes_cpp <- function(Mu, Sigmalist, n, J, nsims) {
    .Call('_test_predict_surbayes_cpp', PACKAGE = 'test', Mu, Sigmalist, n, J, nsims)
}

#' Sample Sigma via Gibbs for SUR model
#' 
#' This is a c++ implementation of sampling Sigma via Gibbs in SUR model--inverse Wishart
#'
#' @param nu degrees of freedom
#' @param V scale matrix
#' @param p dimension of covariance matrix
#' @return sampled covariance matrix
sample_sigma <- function(nu, V, p) {
    .Call('_test_sample_sigma', PACKAGE = 'test', nu, V, p)
}

#' Power Prior Gibbs sampling
#' 
#' This is a c++ implementation of Gibbs sampling SUR model with power prior
#'
#' @param Sigma initial value for covariance matrix
#' @param M number of samples
#' @param X design matrix for current data
#' @param X0 design matrix for historical data
#' @param XtX matrix that is \code{crossprod(cbind(X1, ..., XJ))}
#' @param X0tX0 matrix that is \code{crossprod(cbind(X01, ..., X0J))}
#' @param Y future response as matrix (Y1, ..., YJ)
#' @param Y0 historical response as matrix (Y01, ..., Y0J)
#' @param y future response as vector
#' @param y0 historical response as vector
#' @param a0 power prior parameter
#' @param pvec \code{vector} giving number of covariates per endpoint
#' @param burnin Burn-in parameter
#' @param thin Thin parameter
#' @return sampled covariance matrix
sur_sample_gibbs_cpp <- function(Sigma, M, X, X0, XtX, X0tX0, Y, Y0, y, y0, a0, pvec, burnin, thin) {
    .Call('_test_sur_sample_gibbs_cpp', PACKAGE = 'test', Sigma, M, X, X0, XtX, X0tX0, Y, Y0, y, y0, a0, pvec, burnin, thin)
}

#' Helper function to sample covariance
#' 
#' This function is called by \code{sur_sample_cov_cpp}.
#' It samples the covariance matrix of a SUR
#'
#' @param Y A \code{matrix}, each column a \code{vector} of responses
#' @param Xlist A \code{list}, each element a design \code{matrix}
#' @param n Integer giving number of observations
#' @param J Integer giving number of endpoints
#' @param pj A \code{vector} giving number of covariates per endpoint
#' @param sigma11 A scalar giving a draw for the (1,1) component of the covariance matrix
#' @param r1 A \code{vector} of residuals for the first endpoint's regression
sur_sample_cov_helper_cpp <- function(Y, Xlist, n, J, pj, sigma11, r1) {
    .Call('_test_sur_sample_cov_helper_cpp', PACKAGE = 'test', Y, Xlist, n, J, pj, sigma11, r1)
}

#' Sample from SUR via Direct Monte Carlo (C++ version)
#' 
#' C++ implementation of Zellner and Ando (2010) Direct Monte Carlo
#' method for sampling from the posterior of a Bayesian SUR
#'
#' @param Y \code{matrix} \eqn{(y_1, \ldots y_J)}
#' @param Xlist A \code{list}, each element a design \code{matrix}
#' @param y \code{vector} of responses
#' @param X design \code{matrix}
#' @param XtX \code{matrix} giving \code{crossprod(cbind(X1, ..., XJ))}
#' @param pj \code{vector} giving number of covariates per endpoint
#' @param M An integer giving the number of desired samples
sur_sample_cpp <- function(Y, Xlist, y, X, XtX, pj, M) {
    .Call('_test_sur_sample_cpp', PACKAGE = 'test', Y, Xlist, y, X, XtX, pj, M)
}



R/sur_sample_dmc.R
#' Sample SUR model via direct Monte Carlo
#'
#' This function samples from the posterior of a SUR model using the DMC method of Ando and Zellner (2010)
#' 
#' @importFrom stats as.formula lm model.matrix terms
#' @param formula.list A list of formulas, each element giving the formula for the corresponding endpoint.
#' @param data A \code{data.frame} containing all the variables contained in \code{formula.list]}
#' @param M Number of samples to be drawn
#' @return A list. First element is posterior draws. Second element is list of JxJ covariance matrices. Other elements are helpful statistics about the SUR model to pass to other functions.
#' 
#' @examples
#' ## Taken from bayesm package
#' if(nchar(Sys.getenv("LONG_TEST")) != 0) {M=1000} else {M=10}
#' set.seed(66)
#' ## simulate data from SUR
#' beta1 = c(1,2)
#' beta2 = c(1,-1,-2)
#' nobs = 100
#' nreg = 2
#' iota = c(rep(1, nobs))
#' X1 = cbind(iota, runif(nobs))
#' X2 = cbind(iota, runif(nobs), runif(nobs))
#' Sigma = matrix(c(0.5, 0.2, 0.2, 0.5), ncol = 2)
#' U = chol(Sigma)
#' E = matrix( rnorm( 2 * nobs ), ncol = 2) %*% U
#' y1 = X1 %*% beta1 + E[,1]
#' y2 = X2 %*% beta2 + E[,2]
#' X1 = X1[, -1]
#' X2 = X2[, -1]
#' data = data.frame(y1, y2, X1, X2)
#' names(data) = c( paste0( 'y', 1:2 ), paste0('x', 1:(ncol(data) - 2) ))
#' ## run DMC sampler
#' formula.list = list(y1 ~ x1, y2 ~ x2 + x3)
#' 
#' ## fit using historical data as current data set--never done in practice
#' out = sur_sample_powerprior( formula.list, data, histdata = data, M = M )
#' 
#' @import Rcpp
#' @importFrom Matrix bdiag
#' @importFrom rlist list.cbind
#' @export
sur_sample_dmc <- function(
  formula.list,
  data,
  M = 1000
) {
  
  ##
  ## Perform checks
  ## 
    ## check formula.list
    if ( any( sapply( formula.list, class ) != 'formula' ) ) {
      stop('formula.list must be a list of formulas')
    }
    
    ## check data and histdata
    if ( class(data) != 'data.frame' ) { stop('data must be a data.frame') }
    
    ## Check if M is positive integer
    if ( M <= 0 | M %% 1 != 0 ) { stop('M must be a positive integer') }
  
  
  
  ## Get various needed parameters to pass to functions
  Xlist <- lapply(formula.list, model.matrix, data = data)
  Xnames.list <- lapply(formula.list, function(f) all.vars(f)[-1])  ## names of predictors
  ynames <- sapply(formula.list, function(f) all.vars(f)[1] )       ## names of responses
  Y <- as.matrix(data[, ynames])                    ## matrix of responses (Y1, ..., YJ)
  y <- as.vector(Y)                                 ## vector (y1', ..., yJ')'
  X <- as.matrix(Matrix::bdiag(Xlist))              ## full design matrix
  XtX <- crossprod(rlist::list.cbind(Xlist))        ## p x p matrix of crossproducts
  p.j <- sapply(Xlist, ncol)                        ## number of covars / endpoint
  n <- nrow(data)                                   ## number of observations
  J <- ncol(Y)                                      ## number of endpoints
  
  ## Get names of predictors as j.xname as vector
  pvec <- sapply(Xlist, function(x) ncol(x))
  J <- length(formula.list)
  Xnames <- lapply(formula.list, function(f, data) colnames(model.matrix(f, data)), data = data)
  for(i in 1:J) {
    Xnames[[i]] <- paste0(i, '.', Xnames[[i]])
  }
  Xnames <- unlist(Xnames)
  
  
  ## Call C++ function to sample beta, Sigma | y, X
  sample.list <- sur_sample_cpp(Y, Xlist, y, X, XtX, p.j, M)
  
  ## Name the columns of beta samples
  colnames(sample.list$betadraw) <- Xnames
  
  res <- list(
    'betadraw' = sample.list$betadraw,
    'Sigmalist' = sample.list$covlist,
    'formula.list' = formula.list,
    'ynames' = ynames,
    'Xnames' = Xnames,
    'num.covars' = p.j,
    'n' = n,
    'J' = J,
    'method' = 'dmc'
  )
  ## Set the name for the class
  class(res) <- append(class(res),"surbayes")
  return(res)
}


src/RcppExports.cpp
// Generated by using Rcpp::compileAttributes() -> do not edit by hand
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include <RcppArmadillo.h>
#include <Rcpp.h>

using namespace Rcpp;

// fastKronEye_Y
arma::vec fastKronEye_Y(arma::mat const& Sigma, arma::mat const& Y, int const& n, int const& J);
RcppExport SEXP _test_fastKronEye_Y(SEXP SigmaSEXP, SEXP YSEXP, SEXP nSEXP, SEXP JSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat const& >::type Sigma(SigmaSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type Y(YSEXP);
    Rcpp::traits::input_parameter< int const& >::type n(nSEXP);
    Rcpp::traits::input_parameter< int const& >::type J(JSEXP);
    rcpp_result_gen = Rcpp::wrap(fastKronEye_Y(Sigma, Y, n, J));
    return rcpp_result_gen;
END_RCPP
}
// fastKronEye_crossprod
arma::mat fastKronEye_crossprod(arma::mat const& XtX, arma::mat const& Sigma, arma::vec const& pvec, int const& n, int const& J);
RcppExport SEXP _test_fastKronEye_crossprod(SEXP XtXSEXP, SEXP SigmaSEXP, SEXP pvecSEXP, SEXP nSEXP, SEXP JSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat const& >::type XtX(XtXSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type Sigma(SigmaSEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type pvec(pvecSEXP);
    Rcpp::traits::input_parameter< int const& >::type n(nSEXP);
    Rcpp::traits::input_parameter< int const& >::type J(JSEXP);
    rcpp_result_gen = Rcpp::wrap(fastKronEye_crossprod(XtX, Sigma, pvec, n, J));
    return rcpp_result_gen;
END_RCPP
}
// predict_surbayes_helper
arma::mat predict_surbayes_helper(arma::vec const& mu, arma::mat const& Sigma, int const& n, int const& J);
RcppExport SEXP _test_predict_surbayes_helper(SEXP muSEXP, SEXP SigmaSEXP, SEXP nSEXP, SEXP JSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::vec const& >::type mu(muSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type Sigma(SigmaSEXP);
    Rcpp::traits::input_parameter< int const& >::type n(nSEXP);
    Rcpp::traits::input_parameter< int const& >::type J(JSEXP);
    rcpp_result_gen = Rcpp::wrap(predict_surbayes_helper(mu, Sigma, n, J));
    return rcpp_result_gen;
END_RCPP
}
// predict_surbayes_cpp
arma::cube predict_surbayes_cpp(arma::mat const& Mu, List const& Sigmalist, int const& n, int const& J, int const& nsims);
RcppExport SEXP _test_predict_surbayes_cpp(SEXP MuSEXP, SEXP SigmalistSEXP, SEXP nSEXP, SEXP JSEXP, SEXP nsimsSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat const& >::type Mu(MuSEXP);
    Rcpp::traits::input_parameter< List const& >::type Sigmalist(SigmalistSEXP);
    Rcpp::traits::input_parameter< int const& >::type n(nSEXP);
    Rcpp::traits::input_parameter< int const& >::type J(JSEXP);
    Rcpp::traits::input_parameter< int const& >::type nsims(nsimsSEXP);
    rcpp_result_gen = Rcpp::wrap(predict_surbayes_cpp(Mu, Sigmalist, n, J, nsims));
    return rcpp_result_gen;
END_RCPP
}
// sample_sigma
arma::mat sample_sigma(double const& nu, arma::mat const& V, int const& p);
RcppExport SEXP _test_sample_sigma(SEXP nuSEXP, SEXP VSEXP, SEXP pSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< double const& >::type nu(nuSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type V(VSEXP);
    Rcpp::traits::input_parameter< int const& >::type p(pSEXP);
    rcpp_result_gen = Rcpp::wrap(sample_sigma(nu, V, p));
    return rcpp_result_gen;
END_RCPP
}
// sur_sample_gibbs_cpp
List sur_sample_gibbs_cpp(arma::mat Sigma, int const& M, arma::mat const& X, arma::mat const& X0, arma::mat const& XtX, arma::mat const& X0tX0, arma::mat const& Y, arma::mat const& Y0, arma::vec const& y, arma::vec const& y0, double const& a0, arma::vec const& pvec, int burnin, int thin);
RcppExport SEXP _test_sur_sample_gibbs_cpp(SEXP SigmaSEXP, SEXP MSEXP, SEXP XSEXP, SEXP X0SEXP, SEXP XtXSEXP, SEXP X0tX0SEXP, SEXP YSEXP, SEXP Y0SEXP, SEXP ySEXP, SEXP y0SEXP, SEXP a0SEXP, SEXP pvecSEXP, SEXP burninSEXP, SEXP thinSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat >::type Sigma(SigmaSEXP);
    Rcpp::traits::input_parameter< int const& >::type M(MSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type X(XSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type X0(X0SEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type XtX(XtXSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type X0tX0(X0tX0SEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type Y(YSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type Y0(Y0SEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type y(ySEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type y0(y0SEXP);
    Rcpp::traits::input_parameter< double const& >::type a0(a0SEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type pvec(pvecSEXP);
    Rcpp::traits::input_parameter< int >::type burnin(burninSEXP);
    Rcpp::traits::input_parameter< int >::type thin(thinSEXP);
    rcpp_result_gen = Rcpp::wrap(sur_sample_gibbs_cpp(Sigma, M, X, X0, XtX, X0tX0, Y, Y0, y, y0, a0, pvec, burnin, thin));
    return rcpp_result_gen;
END_RCPP
}
// sur_sample_cov_helper_cpp
arma::mat sur_sample_cov_helper_cpp(arma::mat const& Y, List const& Xlist, int const& n, int const& J, arma::vec const& pj, double const& sigma11, arma::vec const& r1);
RcppExport SEXP _test_sur_sample_cov_helper_cpp(SEXP YSEXP, SEXP XlistSEXP, SEXP nSEXP, SEXP JSEXP, SEXP pjSEXP, SEXP sigma11SEXP, SEXP r1SEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat const& >::type Y(YSEXP);
    Rcpp::traits::input_parameter< List const& >::type Xlist(XlistSEXP);
    Rcpp::traits::input_parameter< int const& >::type n(nSEXP);
    Rcpp::traits::input_parameter< int const& >::type J(JSEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type pj(pjSEXP);
    Rcpp::traits::input_parameter< double const& >::type sigma11(sigma11SEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type r1(r1SEXP);
    rcpp_result_gen = Rcpp::wrap(sur_sample_cov_helper_cpp(Y, Xlist, n, J, pj, sigma11, r1));
    return rcpp_result_gen;
END_RCPP
}
// sur_sample_cpp
List sur_sample_cpp(arma::mat const& Y, List const& Xlist, arma::vec const& y, arma::mat const& X, arma::mat const& XtX, arma::vec const& pj, int const& M);
RcppExport SEXP _test_sur_sample_cpp(SEXP YSEXP, SEXP XlistSEXP, SEXP ySEXP, SEXP XSEXP, SEXP XtXSEXP, SEXP pjSEXP, SEXP MSEXP) {
BEGIN_RCPP
    Rcpp::RObject rcpp_result_gen;
    Rcpp::RNGScope rcpp_rngScope_gen;
    Rcpp::traits::input_parameter< arma::mat const& >::type Y(YSEXP);
    Rcpp::traits::input_parameter< List const& >::type Xlist(XlistSEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type y(ySEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type X(XSEXP);
    Rcpp::traits::input_parameter< arma::mat const& >::type XtX(XtXSEXP);
    Rcpp::traits::input_parameter< arma::vec const& >::type pj(pjSEXP);
    Rcpp::traits::input_parameter< int const& >::type M(MSEXP);
    rcpp_result_gen = Rcpp::wrap(sur_sample_cpp(Y, Xlist, y, X, XtX, pj, M));
    return rcpp_result_gen;
END_RCPP
}

static const R_CallMethodDef CallEntries[] = {
    {"_test_fastKronEye_Y", (DL_FUNC) &_test_fastKronEye_Y, 4},
    {"_test_fastKronEye_crossprod", (DL_FUNC) &_test_fastKronEye_crossprod, 5},
    {"_test_predict_surbayes_helper", (DL_FUNC) &_test_predict_surbayes_helper, 4},
    {"_test_predict_surbayes_cpp", (DL_FUNC) &_test_predict_surbayes_cpp, 5},
    {"_test_sample_sigma", (DL_FUNC) &_test_sample_sigma, 3},
    {"_test_sur_sample_gibbs_cpp", (DL_FUNC) &_test_sur_sample_gibbs_cpp, 14},
    {"_test_sur_sample_cov_helper_cpp", (DL_FUNC) &_test_sur_sample_cov_helper_cpp, 7},
    {"_test_sur_sample_cpp", (DL_FUNC) &_test_sur_sample_cpp, 7},
    {NULL, NULL, 0}
};

RcppExport void R_init_test(DllInfo *dll) {
    R_registerRoutines(dll, NULL, CallEntries, NULL, NULL);
    R_useDynamicSymbols(dll, FALSE);
}


R/sur_sample_powerprior.R
#' Sample from SUR posterior via power prior
#'
#' This function uses Gibbs sampling to sample from the posterior density of a SUR model
#' using the power prior.
#' 
#' @importFrom stats as.formula lm model.matrix terms
#' 
#' @param formula.list A list of formulas, each element giving the formula for the corresponding endpoint.
#' @param data A \code{data.frame} containing all the variables contained in \code{formula.list]}
#' @param histdata A \code{data.frame} of historical data to apply power prior on
#' @param M Number of samples to be drawn
#' @param Sigma0 A \eqn{J \times J} \code{matrix} giving the initial covariance matrix. Default is the MLE.
#' @param a0 A scalar between 0 and 1 giving the power prior parameter
#' @param burnin A non-negative integer giving the burn-in parameter
#' @param thin A positive integer giving the thin parameter
#' 
#' @return A list. First element is posterior draws. Second element is list of JxJ covariance matrices.
#' @examples
#' ## Taken from bayesm package
#' if(nchar(Sys.getenv("LONG_TEST")) != 0) {M=1000} else {M=10}
#' set.seed(66)
#' ## simulate data from SUR
#' beta1 = c(1,2)
#' beta2 = c(1,-1,-2)
#' nobs = 100
#' nreg = 2
#' iota = c(rep(1, nobs))
#' X1 = cbind(iota, runif(nobs))
#' X2 = cbind(iota, runif(nobs), runif(nobs))
#' Sigma = matrix(c(0.5, 0.2, 0.2, 0.5), ncol = 2)
#' U = chol(Sigma)
#' E = matrix( rnorm( 2 * nobs ), ncol = 2) %*% U
#' y1 = X1 %*% beta1 + E[,1]
#' y2 = X2 %*% beta2 + E[,2]
#' X1 = X1[, -1]
#' X2 = X2[, -1]
#' data = data.frame(y1, y2, X1, X2)
#' names(data) = c( paste0( 'y', 1:2 ), paste0('x', 1:(ncol(data) - 2) ))
#' ## run DMC sampler
#' formula.list = list(y1 ~ x1, y2 ~ x2 + x3)
#' 
#' ## fit using historical data as current data set--never done in practice
#' out = sur_sample_powerprior( formula.list, data, histdata = data, M = M )
#' @export
sur_sample_powerprior <- function(
  formula.list, data, histdata, M, Sigma0 = NULL, a0 = 1, burnin = 0, thin = 1
) {
  
  ##
  ## Perform checks
  ## 
    ## check formula.list
    if ( any( sapply( formula.list, class ) != 'formula' ) ) {
      stop('formula.list must be a list of formulas')
    }
    
    ## check data and histdata
    if ( class(data) != 'data.frame' ) { stop('data must be a data.frame') }
    if ( class(histdata) != 'data.frame' ) { stop('histdata must be a data.frame') }
  
    ## Check if M is positive integer
    if ( M <= 0 | M %% 1 != 0 ) { stop('M must be a positive integer') }
  
    ## Make sure Sigma0 is NULL or a matrix of proper dimensions
    if ( !is.null(Sigma0) ) {
      if ( class(Sigma0) != 'matrix' ) {
        stop('Sigma0 must be a matrix')
      }
      if ( nrow(Sigma0) != ncol(Sigma0) ) {
        stop('Sigma0 must be square')
      }
      if ( any(eigen(Sigma0)$values <= 0) ) {
        stop('Sigma0 must be positive definite')
      }
      if ( length(formula.list) != nrow(Sigma0) ) {
        stop('Sigma0 and formula.list must have the same dimensions')
      }
    }
    
    ## Check 0 < a0 <= 1
    if ( a0 <= 0 | a0 > 1 ) { stop('a0 must be larger than 0 and less than or equal to 1') }
  
    ## check burnin is non-negative integer
    if ( !( burnin >= 0 & burnin %% 1 == 0 ) ) {
      stop('burnin must be a non-negative integer')
    }
  
    if ( !(thin >= 1 & burnin %% 1 == 0 ) ) {
      stop('burnin must be a positive integer')
    }
  
  
  ## Number of endpoints / obs
  n <- nrow(data)
  J <- length(formula.list)
  
  ## Helper functions to get Xlist and Y matrix
  get_Xlist <- function(formula.list, data) {
    Xlist = lapply(formula.list, function(f, data) { model.matrix(f, data = data) }, data = data )
  }
  get_Ymat <- function(formula.list, data) {
    ynames <- sapply(formula.list, function(f, data) all.vars(f)[1])
    Ymat <- as.matrix(data[, ynames])
  }
  
  ## Obtain list of X's and extract names
  Xlist0 <- get_Xlist(formula.list, histdata)
  Xlist <- get_Xlist(formula.list, data)
  ynames <- sapply(formula.list, function(f) all.vars(f)[1] )       ## names of responses
  

  ## Get names of predictors as j.xname as vector
  pvec <- sapply(Xlist, function(x) ncol(x))
  J <- length(formula.list)
  Xnames <- lapply(formula.list, function(f, data) colnames(model.matrix(f, data)), data = data)
  for(i in 1:J) {
    Xnames[[i]] <- paste0(i, '.', Xnames[[i]])
  }
  Xnames <- unlist(Xnames)
  
  
  ## Get response matrix (Y1, ..., YJ)
  Y0 <- get_Ymat(formula.list, histdata)
  Y <- get_Ymat(formula.list, data)
  y0 <- as.vector(Y0)
  y <- as.vector(Y)
  
  X0 <- as.matrix(Matrix::bdiag(Xlist0))
  X <- as.matrix(Matrix::bdiag(Xlist))
  XtX <- crossprod(rlist::list.cbind(Xlist))
  X0tX0 <- crossprod(rlist::list.cbind(Xlist0))
  
  if ( is.null(Sigma0) ) {
    resmat <- sapply(formula.list, function(f, data) stats::resid(lm(f, data) ), data = data )
    Sigma0 <- crossprod(resmat) / nrow(data)
  }
  
  sample <- sur_sample_gibbs_cpp( Sigma0, M, X, X0, XtX, X0tX0, Y, Y0, y, y0, a0, pvec, burnin, thin )
  
  betasample <- sample$betasample
  colnames(betasample) <- Xnames
  Sigmalist <- sample$Sigmalist
  
  
  res <- list(
    'betadraw' = betasample,
    'Sigmalist' = Sigmalist,
    'formula.list' = formula.list,
    'ynames' = ynames,
    'Xnames' = Xnames,
    'num.covars' = pvec,
    'n' = n,
    'J' = J,
    'method' = 'gibbs'
  )
  
  ## Set the name for the class
  class(res) <- append(class(res), "surbayes" )
  return(res)
}

